### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by Marcinati**
1. InvalidGear exception
2. Engines as a unique_ptr
3. Encapsulate members
4. InvalidSpeed exception
5. Change the engine method
6. Charging of electric car
7. Refueling of petrol car
8. Decelerate speed method
9. Changing gears of petrol car
10. Invalid Restore when capacity/tank is full
11. Invalid DriveMode of hybrid car
12. Changing DriveMode of hybird
13. Tests for all car types
14. Workflow file with Valgrind

### CODE REVIEW

File: Car.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include "Exceptions.hpp"
+
+void Car::turnLeft() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+void Car::turnRight() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+void Car::accelerate(int speed) {
+    if (speed < 0) {
```
Comment (MaciejSalamonski) [inline Car.cpp:12]: You often usibg 0 value. Lets make zero as constexpr value.

**Review by MaciejSalamonski — APPROVED** [review]
Hi. Nice code :) have a great day!

File: ElectricEngine.hpp [hunk 2]
```diff
@@ -1,11 +1,10 @@
...
+class ElectricEngine {
 public:
-    ElectricEngine(int power, int batteryCapacity);
+    ElectricEngine(int power, int maxBatteryCapacity, int currentBatteryCapacity);
 
-    int power_;             // in HP 
-    int batteryCapacity_;   // in Ah
+    int power_;                   // in HP
+    int maxBatteryCapacity_;      // in Ah
+    int currentBatteryCapacity_;  //in Ah
```
Comment (ziobron) [inline ElectricEngine.hpp:9]: could be private

File: PetrolEngine.hpp [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+    int maxFuelTank_;      //in l
+    int currentFuelTank_;  //in l
+    std::map<int, std::string> gearBoxMap_ {{-1, "rear"},
+                                           {0, "neutral"},
+                                           {1, "first"},
+                                           {2, "second"},
+                                           {3, "third"},
+                                           {4, "fourth"},
+                                           {5, "fifth"},
+                                           {6, "sixth"}};
```
Comment (ziobron) [inline PetrolEngine.hpp:22]: can be private

File: HybridCar.cpp [hunk 4]
```diff
@@ -1,19 +1,29 @@
...
+void HybridCar::restore() {
+    ElectricCar::restore();
+    PetrolCar::restore();
+}
+void HybridCar::changeEngine(std::unique_ptr<ElectricEngine> newElectricEngine,
+                             std::unique_ptr<PetrolEngine> newPetrolEngine) {
+    std::swap(electricEngine_, newElectricEngine);
+    std::swap(petrolEngine_, newPetrolEngine);
+}
+void HybridCar::changeDriveMode(int mode) {
```
Comment (ziobron) [inline HybridCar.cpp:22]: param could be enum

File: HybridCar-ut.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+TEST_F(HybridTestCar, HybridCarCANTChangElectricEngineWhileItIstMoving) {
+    hybridCar.accelerate(100);
+    EXPECT_THROW(hybridCar.changeElectricEngine(std::make_unique<ElectricEngine>(newEnginePPower,
+                                                                                 newMaxBatteryCapacity,
+                                                                                 newCurrentBatteryCapacity)),
+                                                                                 InvalidSpeed);
+}
+
+TEST_F(HybridTestCar, HybridCarCanChargeWhenHasLowerBatteryCapacityThanengineCapacity) {
+    EXPECT_NO_THROW(hybridCar.ElectricCar::restore());
```
Comment (ziobron) [inline HybridCar-ut.cpp:88]: what about calling just `hybridCar.restore()`?

**Review by ziobron — APPROVED** [review]
Good piece of code 👍

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
9 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**